File: 0343-integer-break/0343-integer-break.py
class Solution:
    def integerBreak(self, n: int) -> int:
        dp = [0 for _ in range(n + 1)]
        dp[2] = 1

        for i in range(3, n + 1):
            for j in range(1, i):
                dp[i] = max(dp[i], dp[i - j] * j, (i - j) * j)

        return dp[n]

File: 0343-integer-break/0343-integer-break.js
/**
 * @param {number} n
 * @return {number}
 */
var integerBreak = function(n) {
    const dp = new Array(n + 1).fill(0);
    
    dp[2] = 1;
    
    for (let i = 3; i <= n; i++) {
        for (let j = 1; j <= i - j; j++) {
            dp[i] = Math.max(Math.max(j * (i - j), dp[i - j] * j), dp[i]);
        }
    }
    
    return dp[n];
};